File: AEDS-II/verde/TP02/Show.java
import java.util.*;
import java.io.*;
import java.text.*;

public class Show{

    static SimpleDateFormat ddf = new SimpleDateFormat("MMMM d, yyyy", Locale.ENGLISH);

    private String SHOW_ID;
    private String TYPE;
    private String TITLE;
    private String[] DIRECTOR;
    private String[] CAST;
    private String COUNTRY;
    private String DATE_ADDED;
    private int RELEASE_YEAR;
    private String RATING;
    private String DURATION;
    private String[] LISTED_IN;

    public Show(){
        this.SHOW_ID = "NaN";
        this.TYPE = "NaN";
        this.TITLE = "NaN";
        this.DIRECTOR = new String[0];
        this.CAST = new String[0];
        this.COUNTRY = "NaN";
        this.DATE_ADDED = "NaN";
        this.RELEASE_YEAR = -1;
        this.RATING = "NaN";
        this.DURATION = "NaN";
        this.LISTED_IN = new String[0];
    }

    public Show(String SHOW_ID, String TYPE, String TITLE, String[] DIRECTOR, String[] CAST,
                String COUNTRY, String DATE_ADDED, int RELEASE_YEAR, String RATING, String DURATION, String[] LISTED_IN){
        this.SHOW_ID = SHOW_ID;
        this.TYPE = TYPE;
        this.TITLE = TITLE;
        this.DIRECTOR = DIRECTOR;
        this.CAST = CAST;
        this.COUNTRY = COUNTRY;
        this.DATE_ADDED = DATE_ADDED;
        this.RELEASE_YEAR = RELEASE_YEAR;
        this.RATING = RATING;
        this.DURATION = DURATION;
        this.LISTED_IN = LISTED_IN;
    }

    public String getSHOW_ID(){
        return this.SHOW_ID;
    }

    public void print(){
        Arrays.sort(DIRECTOR);
        Arrays.sort(CAST);
        Arrays.sort(LISTED_IN);

        String diretorStr = (DIRECTOR.length == 0) ? "NaN" : String.join(", ", DIRECTOR);
        String castStr = (CAST.length == 0) ? "NaN" : String.join(", ", CAST);
        String listedInStr = (LISTED_IN.length == 0) ? "NaN" : String.join(", ", LISTED_IN);

        System.out.println("=> " + (SHOW_ID != null ? SHOW_ID : "NaN") + " ## " +
                (TITLE != null ? TITLE : "NaN") + " ## " +
                (TYPE != null ? TYPE : "NaN") + " ## " +
                diretorStr + " ## " +
                "[" + castStr + "] ## " +
                (COUNTRY != null ? COUNTRY : "NaN") + " ## " +
                (DATE_ADDED != null ? DATE_ADDED : "NaN") + " ## " +
                (RELEASE_YEAR != -1 ? RELEASE_YEAR : "NaN") + " ## " +
                (RATING != null ? RATING : "NaN") + " ## " +
                (DURATION != null ? DURATION : "NaN") + " ## " +
                "[" + listedInStr + "] ##");
    }

    public static Show parseLinha(String linha){
        String[] campos = new String[12];
        int index = 0;
        boolean dentroAspas = false;
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < linha.length(); i++){
            char c = linha.charAt(i);

            if (c == '"'){
                dentroAspas = !dentroAspas;
            } else if (c == ',' && !dentroAspas){
                campos[index++] = sb.toString().trim();
                sb.setLength(0);
            } else{
                sb.append(c);
            }
        }
        campos[index] = sb.toString().trim();

        String[] dir = campos[3] != null && !campos[3].isEmpty() ? campos[3].split(", ") : new String[0];
        String[] cast = campos[4] != null && !campos[4].isEmpty() ? campos[4].split(", ") : new String[0];
        String[] listed = campos[10] != null && !campos[10].isEmpty() ? campos[10].split(", ") : new String[0];

        int year = -1;
        try{ year = Integer.parseInt(campos[7]); } catch (Exception ignored){}

        return new Show(
                campos[0], campos[1], campos[2],
                dir, cast,
                campos[5],
                campos[6],
                year,
                campos[8],
                campos[9],
                listed
        );
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        String entrada = sc.nextLine();

        while (!entrada.equals("FIM")){
            boolean encontrado = false;

            try (BufferedReader br = new BufferedReader(new FileReader("/home/vicenzo/Documentos/GitHub/Lista-AEDS/AEDS-II/verde/TP02/disneyplus.csv"))){
                br.readLine(); // pular cabeçalho
                String linha;

                while ((linha = br.readLine()) != null){
                    Show show = parseLinha(linha);

                    if (show.getSHOW_ID().equals(entrada)){
                        show.print();
                        encontrado = true;
                        break;
                    }
                }
            } catch (IOException e){
                e.printStackTrace();
            }

            if (!encontrado){
                System.out.println("Show não encontrado.");
            }

            entrada = sc.nextLine();
        }

        sc.close();
    }
}

File: AEDS-II/verde/TP02/C/Show.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define TAM_LINHA 1024
#define MAX_CAMPOS 12
#define TAM_MAX_LISTA 100
#define MAX_SHOWS 5000

typedef struct{
    char *SHOW_ID;
    char *TYPE;
    char *TITLE;
    char **DIRECTOR;
    int DIRECTOR_COUNT;
    char **CAST;
    int CAST_COUNT;
    char *COUNTRY;
    char *DATE_ADDED;
    int RELEASE_YEAR;
    char *RATING;
    char *DURATION;
    char **LISTED_IN;
    int LISTED_COUNT;
} Show;

char *CLONE(const char *s){
    char *copy = malloc(strlen(s) + 1);
    if (copy) strcpy(copy, s);
    return copy;
}

char **split_lista(const char *str, int *qtd){
    char **lista = malloc(TAM_MAX_LISTA * sizeof(char *));
    *qtd = 0;
    char *aux = CLONE(str);
    char *token = strtok(aux, ",");
    while (token != NULL && *qtd < TAM_MAX_LISTA){
        while (*token == ' ') token++;
        lista[(*qtd)++] = CLONE(token);
        token = strtok(NULL, ",");
    }
    free(aux);
    return lista;
}

void ordenar_lista(char **lista, int tamanho){
    for (int i = 0; i < tamanho - 1; i++){
        for (int j = 0; j < tamanho - i - 1; j++){
            if (strcmp(lista[j], lista[j + 1]) > 0){
                char *temp = lista[j];
                lista[j] = lista[j + 1];
                lista[j + 1] = temp;
            }
        }
    }
}

int LER_LINHA(char *linha, Show *s){
    char *campos[MAX_CAMPOS];
    int index_campos = 0;
    char buffer[TAM_LINHA];
    int index_buffer = 0;
    int entreAspas = 0;

    for (int i = 0; linha[i] != '\0'; i++){
        if (linha[i] == '"'){
            entreAspas = !entreAspas;
        } else if (linha[i] == ',' && !entreAspas){
            buffer[index_buffer] = '\0';
            campos[index_campos] = CLONE(buffer);
            index_buffer = 0;
            index_campos++;
            if (index_campos >= MAX_CAMPOS) break;
        } else{
            buffer[index_buffer++] = linha[i];
        }
    }
    buffer[index_buffer] = '\0';
    campos[index_campos++] = CLONE(buffer);

    if (index_campos < 11) return 0;

    s -> SHOW_ID = campos[0];
    s -> TYPE = campos[1];
    s -> TITLE = campos[2];

    s -> DIRECTOR = split_lista(campos[3], &s -> DIRECTOR_COUNT);
    ordenar_lista(s -> DIRECTOR, s -> DIRECTOR_COUNT);

    s -> CAST = split_lista(campos[4], &s -> CAST_COUNT);
    ordenar_lista(s -> CAST, s -> CAST_COUNT);

    s -> COUNTRY = campos[5];
    s -> DATE_ADDED = campos[6];
    s -> RELEASE_YEAR = atoi(campos[7]);
    s -> RATING = campos[8];
    s -> DURATION = campos[9];

    s -> LISTED_IN = split_lista(campos[10], &s -> LISTED_COUNT);
    ordenar_lista(s -> LISTED_IN, s -> LISTED_COUNT);

    return 1;
}

void print_lista(char **lista, int qtd){
    if (qtd == 0){
        printf("NaN");
    } else{
        for (int i = 0; i < qtd; i++){
            if (i > 0) printf(", ");
            printf("%s", lista[i]);
        }
    }
}

void print_show(Show *s){
    printf("=> %s ## %s ## %s ## ", s -> SHOW_ID, s -> TITLE, s -> TYPE);
    print_lista(s -> DIRECTOR, s -> DIRECTOR_COUNT);
    printf(" ## [");
    print_lista(s -> CAST, s -> CAST_COUNT);
    printf("] ## %s ## %s ## ", s -> COUNTRY, s -> DATE_ADDED);

    if (s -> RELEASE_YEAR != -1)
        printf("%d ## ", s -> RELEASE_YEAR);
    else
        printf("NaN ## ");

    printf("%s ## %s ## [", s -> RATING, s -> DURATION);
    print_lista(s -> LISTED_IN, s -> LISTED_COUNT);
    printf("] ##\n");
}

int main(){
    FILE *arq = fopen("/tmp/disneyplus.csv", "r");

    if (!arq){
        perror("Erro ao abrir o arquivo");
        return 1;
    }

    Show shows[MAX_SHOWS];
    int show_qtd = 0;
    char linha[TAM_LINHA];

    fgets(linha, TAM_LINHA, arq); // Pular cabeçalho

    while (fgets(linha, TAM_LINHA, arq) && show_qtd < MAX_SHOWS){
        if (LER_LINHA(linha, &shows[show_qtd])){
            show_qtd++;
        }
    }
    fclose(arq);

    char entrada[20];
    scanf("%s", entrada);

    while (strcmp(entrada, "FIM") != 0){
        int found = 0;
        for (int i = 0; i < show_qtd; i++){
            if (strcmp(shows[i].SHOW_ID, entrada) == 0){
                print_show(&shows[i]);
                found = 1;
                break;
            }
        }
        if (!found) printf("Show n\xC3\xA3o encontrado.\n");
        scanf("%s", entrada);
    }

    return 0;
}
